var/exp_repo/experiments/amg2023/experiment.py

```diff
@@ -23,173 +21,111 @@ class Amg2023(Experiment):
         description="type of experiment",
     )
 
-    def make_experiment_example(self):
-        app_name = self.spec.name
+    variant(
+        "version",
+        default="develop",
+        description="app version",
+    )
 
-        variables = {}
-        matrices = []
-        zips = {}
+    # requires("system+papi", when(caliper=topdown*))
 
-        if self.spec.satisfies("programming_model=openmp"):
+    # TODO: Support list of 3-tuples
+    # variant(
+    #     "p",
+    #     description="value of p",
+    # )
+
+    # TODO: Support list of 3-tuples
+    # variant(
+    #     "n",
+    #     description="value of n",
+    # )
+
+    def make_experiment_example(self):
+        self.add_experiment_name_prefix("example")
+
+        if self.spec.satisfies("openmp=oui"):
             # TODO: Support variants
             n = ["55", "110"]
-            variables["n_nodes"] = ["1", "2"]
-            variables["n_ranks"] = "8"
-            variables["n_threads_per_proc"] = ["4", "6", "12"]
-            exp_name = f"{app_name}_example_omp_{{n_nodes}}_{{n_ranks}}_{{n_threads_per_proc}}_{{px}}_{{py}}_{{pz}}_{{nx}}_{{ny}}_{{nz}}"
-        elif self.spec.satisfies("programming_model=cuda"):
+            self.add_experiment_variable("n_nodes", ["1", "2"], True)
+            self.add_experiment_variable("n_ranks", "8", True)
+            self.add_experiment_variable("n_threads_per_proc", ["4", "6", "12"], True)
+        elif self.spec.satisfies("cuda=oui"):
             # TODO: Support variants
             n = ["10", "20"]
-            variables["n_gpus"] = "8"
-            exp_name = f"{app_name}_example_cuda_{{n_gpus}}_{{px}}_{{py}}_{{pz}}_{{nx}}_{{ny}}_{{nz}}"
-        elif self.spec.satisfies("programming_model=rocm"):
+            self.add_experiment_variable("n_gpus", "8", True)
+        elif self.spec.satisfies("rocm=oui"):
             # TODO: Support variants
             n = ["110", "220"]
-            variables["n_gpus"] = "8"
-            exp_name = f"{app_name}_example_rocm_{{n_gpus}}_{{px}}_{{py}}_{{pz}}_{{nx}}_{{ny}}_{{nz}}"
+            self.add_experiment_variable("n_gpus", "8", True)
         else:
             raise NotImplementedError(
                 "Unsupported programming_model. Only openmp, cuda and rocm are supported"
             )
 
         # TODO: Support variant
         p = "2"
-        variables["px"] = p
-        variables["py"] = p
-        variables["pz"] = p
+        self.add_experiment_variable("px", p, True)
+        self.add_experiment_variable("py", p, True)
+        self.add_experiment_variable("pz", p, True)
```

Comment:
@scheibelp @pearce8 The application can use an API to set experiment variables and other constructs